src/raw_node.rs

```diff
@@ -263,6 +263,11 @@ impl<T: Storage> RawNode<T> {
         if !rd.read_states.is_empty() {
             self.raft.read_states.clear();
         }
+        // update raft uncommitted entries size
+        if rd.committed_entries.is_some() {
+            self.raft
+                .reduce_uncommitted_size(&rd.committed_entries.unwrap_or_default())
+        }
```

Comment:
Unlike etcd, raft-rs has interfaces like `ready_since` which fetch committed entries in range `(user_provided_index, commit_index]`. In the usage of TiKV, the `user_provided_index` is the last committed entry of the last `Ready` it got [code](https://github.com/tikv/tikv/blob/0b05a9947ca1bc29e090a24a58d8838f1574f9f8/components/raftstore/src/store/peer.rs#L1718). And `reduce_uncommitted_size` for `Ready.committed_entries` here may make  `uncommitted_size` inaccurate, even we compute `uncommitted_size` based on the commit index after a peer become leader, because the `user_provided_index` may less than the commit index

Reply:
A workaround to fix this issue is we compute the `uncommitted_size` again if we meet `uncommitted_size < 0` in `reduce_uncommitted_size`

Reply:
You mean that when using `ready_since`, user may `advance` a `ready` which has entries that has been included in last `ready`?

Reply:
No, for example, a peer is a follower and has log entries: [i1, .., c1, i2, .., c2] 
1. the peer's commit index is c1, so it produces a Ready with entries [i1, .., c1]
2. the peer's commit index forward to c2
3. the peer becomes the leader and set `uncommitted_size` to zero because there are not uncommitted entries
4. the peer advance the previous Ready with entries [i1, .., c1], and call `reduce_uncommitted_size` then find `uncommitted_size` < 0

With the abovementioned workaround, the peer set `uncommitted_size` to the size of entries [i2, .., c2], because next time the peer will call `ready_since(c1)` and produce a Ready with entries [i2, .., c2], and later advance it.

Reply:
What about use an `apply_hint` to keep track of the commit index when a peer becomes leader?

When `reduce_uncommitted_size`, we only need to reduce the size of entry slice from 
`[max(apply_hint+1, first_entry_index), last_entry_index]` 

eg.
For peer which is leader(normal case):
`apply_hint` should always smaller than `first_entry_index` in any `Ready`;

For peer become leader(which is the case we discuss about):
`apply_hint` is set to `raft_log.committed` when peer becomes leader. For any `Ready` created when peer is follower,
`apply_hint` should be greater than the `last_entry_index` in `Ready`. So the size of slice shall be 0 and current
`uncommitted_size` will not be reduced. 

We can update `apply_hint` in `compute_uncommitted_size`

Reply:
Good idea! I have a few suggestions:
- the name `apply_hint` may be ambiguous, I suggest something like `init_committed`
- we can update `apply_hint` in `become_leader`
- we should push `max_uncommitted_size`, `uncommitted_size` and `apply_hint` into a struct to make the logic clear